# Creating a function

# 0. Libraries

In [1]:
import yfinance as yf
import seaborn as sns
import pandas as pd
import numpy as np

%matplotlib inline
import matplotlib.pyplot as plt
sns.set(style="darkgrid")

# 1. Creating a function

This function will do everything what we did before

In [4]:
def test_strategy(stock, start, end, SMA):
    df = yf.download(stock, start = start, end = end)
    data = df['Close'].to_frame()
    data['returns'] = np.log(data['Close'].div(data['Close'].shift(1)))
    data['SMA_S'] = data['Close'].rolling(int(SMA[0])).mean()
    data['SMA_L'] = data['Close'].rolling(int(SMA[1])).mean()
    data.dropna(inplace = True)
    
    data['position'] = np.where(data['SMA_S'] > data['SMA_L'], 1, -1)
    data['strategy'] = data['returns'] * data['position'].shift(1)
    data.dropna(inplace = True)
    
    # absolute performance of the strategy (if 1 dollar is invested, what happens)
    ret = np.exp(data['strategy'].sum())
    std = data['strategy'].std() * np.sqrt(252)
    
    return ret, std

In [6]:
# now we use the function, lets try with microsoft
test_strategy('MSFT', '2000-01-01', '2020-01-01', (50, 200))

[*********************100%%**********************]  1 of 1 completed


(0.39097954932295986, 0.2906766831415829)

We see that this is a very poor return for the SMA since $1 became $0.3908 (ret) and the standard deviation is 29%

In [7]:
# now try for apple:
test_strategy('AAPL', '2000-01-01', '2020-01-01', (50, 200))

[*********************100%%**********************]  1 of 1 completed


(16.540320732436786, 0.36699682903803943)

We see this is much better - $1 became $16.5 (ret) with a standard deviation of 36.7%

In [8]:
# try for tesla
test_strategy('TSLA', '2000-01-01', '2020-01-01', (50, 200))

[*********************100%%**********************]  1 of 1 completed


(1.3908153866370587, 0.4956456904999671)

This is not good since $1 only became $1.39 with a standard deviation of 49.6%

In [9]:
# lets check for SPY
test_strategy('SPY', '2000-01-01', '2020-01-01', (50, 200))

[*********************100%%**********************]  1 of 1 completed


(4.766371508372407, 0.18787947185958592)

$1 became $4.77 with a std of 18.8%